records/2023-01-26.bounce-rate-automated-emergency-brake.md

```diff
@@ -31,47 +33,99 @@ AWS calculates the bounce rate as the number of hard bounces per the total numbe
+####  Provide bounce rate guidance
+Provide guidance to service owners to explain bounce rate concepts and how they apply to their service.  Provide guidance on how services can keep their bounce rate low. 
 
-## Staggered roll-out
-To prevent suspending services that may not actually be causing issues with the overall bounce rate, it is suggested to roll out this feature in stages.  The first stage would include tracking bounces, calculating bounce rates, and displaying them on service dashboards. It would also include bounce rate guidance for services. Once we are sure our calculations work satisfactorily with live data, we would move to the second stage which will include notifying services when they are approaching the bounce rate threshold and suspending services when they reach the threshold.
+####  Bounce rate calculation for a service
+The formula for calculating a service's bounce rate will be the total number of hard bounces the service received divided by the total number of emails the service sent during the `[bounce-rate-window]`.  
+
+### Technical considerations
+#### Changes to the Notification model
+In order to store the bounce rate GC Notify needs to record the bounce type of every notification.  This will require a new column in the notifications table to store the bounce type.  
+
+#### Calculating the bounce rate
+Since the bounce rate can nearly be calculated as-is from the data currently pulled live into the dashboard, it is possible to calculate the bounce rate on the fly in the dashboard with minimal code changes.  This avoids the need to explicitly store the bounce rate in the database or redis.
+
+#### Changes to the dashboard
+The dashboard already pulls weekly and daily sending statistics including sending totals and total failures.  This could be extended to pull data also by `[bounce-rate-window]` to enable bounce rate calculation.  Failure data will need to be broken down further than it currently is to include the bounce type.  
+
+#### Monitoring bounce rates
+Bounce rates can be monitored by a periodic celery task that polls the database for bounce rates, and notifies services that are over `[bounce-rate-warning-percentage]` and suspends services that are over `[bounce-rate-critical-percentage]`.  This task should be run several times a day to ensure that services are notified and suspended as soon as possible after they reach one of these limits.  A query to get service bounces rates could look something like this:
```

Comment:
Makes sense to check bounce rates with a repeating Celery task! 

What about checking each services bounce rate from Redis instead of the database? This would reduce the query load on the database, as we'd need to run this aggregate query often (the more the better, i.e. every minute).

Reply:
We certainly could.  I wasn't really worrying about the DB load since the admin dashboard runs 10 statistics queries  every 5 seconds when a user is on the dashboard, so I figured it was fast enough.

I can update the ADR to include a redis implementation.  Maybe we can centralize that code inside one of the email sending methods (i.e. `send_to_providers.py/send_email_to_provider` to avoid having to track it in 4 places like the sms limit does?

We would also then need to track the number of hard bounces in redis; we'll need to track the synchronous ones in `clients/email/aws_ses.py/send_email()` and the async ones in the `process-ses-result` celery task.  Make sense?

Reply:
I'd avoid repetitive reads on the database if possible. There were cases where we brought down the admin UI because of the DB repetitive reads. I tried to find the old bug card about it (in Trello), but in general that would be a fitting task for a cache server.

Reply:
Sounds good.  As an aside, we may want to look into the dashboard because as I mentioned it does multiple queries every 5 seconds whenever someone has a dashboard open.